### ISSUE TITLE
Will this work with an SSL MITM Proxy?

### ISSUE THREAD

**Issue by matthew-richard** [proxytunnel/proxytunnel#49]
I've been able to set this up with a local proxy that does [SSL Inspection](https://help.zscaler.com/zia/about-ssl-inspection), but I can't seem to establish an SSH connection with my home machine. The local proxy (run by my organization) effectively performs a man in the middle attack on every HTTPS CONNECT request that it receives, meaning it's able to inspect every packet sent over SSL.

I've set up netcat to listen on my home machine `nc -lv -p 443` and then connected to it via proxytunnel on my work machine `proxytunnel -v -p work.proxy.com:8080 -r my.proxy.com:443 -d my.homemachine.com:443`. With that setup, I've been able to type on my work machine and see the messages on my home machine. So, clearly a tunnel is being established.

If I then attempt to SSH into my home machine using this same command (I have SSH set up on port 443 just in case it helps), `ssh -vvv -o "ProxyCommand=proxytunnel -v -p work.proxy.com:8080 -r my.proxy.com:443 -d %h:%p -H 'User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Win32)'" -p 443 user@my.homemachine.com`, I see the initial SSH protocol messages arrive, but then the connection is closed by the local proxy (Output below).

My theory is that the local proxy is inspecting the sent packets, detecting whether the messages resemble an SSH handshake, and closing the connection if they do. That's the only way I can explain being able to establish a tunnel and send arbitrary text over it, while still seeing these SSH connections being closed.

My question being: **Can proxytunnel be thwarted by MITM, SSL inspection proxy servers?** If so, is there another tool that works around this?

```
OpenSSH_for_Windows_7.6p1, LibreSSL 2.6.4
debug3: Failed to open file:C:\\Users\\myuser/.ssh/config error:2
debug3: Failed to open file:C:\\ProgramData\\ssh/ssh_config error:2
debug1: Executing proxy command: exec Cproxytunnel.exe -v -p work.proxy.com:8080 -r my.proxy.com:443 -d my.homemachine.com:443 -H 'User-Agent:  Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/11.11.11.11 Safari/537.36'

debug3: spawning proxytunnel.exe -v -p work.proxy.com:8080 -r my.proxy.com:443 -d my.homemachine.com:22 -H 'User-Agent:  Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/11.11.11.11 Safari/537.36'
debug3: w32_getpeername ERROR: not sock :2
debug1: identity file C:\\Users\\myuser/.ssh/id_rsa type 0
[Some messages about identify files - not relevant]
debug1: identity file C:\\Users\\myuser/.ssh/id_ed25519-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_for_Windows_7.6
Local proxy work.proxy.com resolves to 11.11.11.11
Connected to work.proxy.com:8080 (local proxy)

Tunneling to my.proxy.com:443 (remote proxy)
Communication with local proxy:
 -> CONNECT my.proxy.com:443 HTTP/1.0
 -> Proxy-Connection: Keep-Alive
 -> User-Agent:  Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/11.11.11.11 Safari/537.36
 <- HTTP/1.1 200 Connection Established
 <- Proxy-Agent: Zscaler/6.0

Tunneling to my.homemachine.com:443 (destination)
Communication with remote proxy:
 -> CONNECT my.homemachine.com:443 HTTP/1.0
 -> Proxy-Connection: Keep-Alive
 -> User-Agent:  Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/11.11.11.11 Safari/537.36
 <- HTTP/1.1 200 Connection Established
 <- Proxy-Agent: Zscaler/6.0

Tunnel established.
debug1: Remote protocol version 2.0, remote software version OpenSSH_for_Windows_7.7
debug1: match: OpenSSH_for_Windows_7.7 pat OpenSSH* compat 0x04000000
debug2: fd 5 setting O_NONBLOCK
debug2: fd 4 setting O_NONBLOCK
debug1: Authenticating to my.homemachine.com:22 as 'MY-HOMEMACHINE\\myuser'
debug3: hostkeys_foreach: reading file "C:\\Users\\myuser/.ssh/known_hosts"
debug3: Failed to open file:C:\\Users\\myuser/.ssh/known_hosts2 error:2
debug3: Failed to open file:C:\\ProgramData\\ssh/ssh_known_hosts error:2
debug3: Failed to open file:C:\\ProgramData\\ssh/ssh_known_hosts2 error:2
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
Tunnel closed.
```

**Comment by sigio** [comment]
On Mon, 21 Dec 2020 09:24:03 -0800
Matt Richard <notifications@github.com> wrote:

> My question being: **Can proxytunnel be thwarted by MITM, SSL inspection proxy servers?** If so, is there another tool that works around this?

That's the entire goal of MITM-ing proxies.... so yeah. Proxytunnel is
more then 15 years old, and was mostly written for the time that
web-proxies might have inspected regular http traffic, but left CONNECT
requests completely alone.

If your proxy detects SSH or something not completely like http(s) it
will probably block it, if configured to do so.

-- 
Mark Janssen

**Comment by matthew-richard** [comment]
Hm, thanks for your response.

What surprises me is that there isn't some tool out there to tunnel one layer deeper. I.e. send actual HTTP requests, with HTML bodies, but in the HTML include (base64 encoded) messages that are then decoded (and/or decrypted) by a server side binary, then proxied from there as usual. AFAIK the MITM-ing proxies would have no way of distinguishing that from legitimate web browsing. 

Or, more simply - Since apparently the MITM proxy doesn't care if what it's seeing is HTML, as long as it's not SSH handshake messages - Encrypt (using some basic symmetric encryption scheme) every message sent over SSL, so that looks like junk to the MITM, then decrypt it on the server side. The symmetric encryption can be weak - It just has to fool the MITM, and then a more secure SSL connection can be established within it.

Not saying I plan on writing such a tool! But, it sounds completely feasible and I'm surprised there isn't someone as frustrated as me out there who's written it already. As it is now, I have no secure way of talking to my home machine aside from using GPG to encrypt stuff before sending it over my MITM-d proxytunnel connection.

In general I don't understand aggressive security protocols like this. It just seems like a matter of time until someone finds a workaround, since every system has an internet connection, even if it's through a proxy.

**Comment by FlexMcMurphy** [comment]
Did you try using proxytunnels -e parameter? This command can get through a work proxy... I don't know about proxies that can do a MITM inspection.

`proxytunnel -v -e -p work-proxy-ip:port -N -P my-username:my-password -t my-work-domain -d my-home-router-public-ip:443 -H "User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Win32)"`

Inside your home you could use sslh and stunnel to decrypt the tls encrypted ssh communication and then forward it to your ssh server.

Alternatively, you could try a VPN to tunnel out?

Flex

**Comment by matthew-richard** [comment]
I've experimented with the -e, -E, and -X parameters. No luck.

I do think VPN would be the next thing to try here, although I'm not sure if that uses SSL for encrypting communications too, in which case the proxy could pull another MITM. It's just a matter of whether the proxy, looking at the unencrypted comms, already has some policy for rejecting what "looks like" SSH or VPN communications. 

At least, that's what I suspect, given that I can open a raw socket connection with my home machine. 

Unfortunately I've already spent a bunch of time on this and don't think I can justify trying another route. I can always just set up a server (over https or http, doesnt make a difference) that stores whatever encrypted data I pass to it. 

Wish I had more time. This has been an interesting technical challenge.
